### ISSUE TITLE
[BUG] Chromium zombie processes when ran inside docker

### ISSUE THREAD

**Issue by Tristan971** [microsoft/playwright-java#876]
**Context:**
- Playwright Version: 1.20.1
- Operating System:
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

- Browser: Chromium (at least)
- Extra: Driver and browsers are preinstalled and cached inside the docker image

**Code Snippet**

There are two parts to this, first the docker image (the official one is really massive since it bundles all browsers, and also is on a JDK 11 baseline instead of JDK 17, hence this one which is a little slimmer).

```Dockerfile
FROM docker.io/library/eclipse-temurin:17

ENV DEBIAN_FRONTEND "noninteractive"
ENV TZ "UTC"

ARG PLAYWRIGHT_BROWSERS_PATH="/ms-playwright"
ENV PLAYWRIGHT_BROWSERS_PATH /ms-playwright

RUN apt-get update && \
    apt-get install -y curl wget && \
    curl -sL https://deb.nodesource.com/setup_16.x | bash - && \
    apt-get install -y nodejs && \
    mkdir /ms-playwright && \
    npx --yes playwright install chromium && \
    npx --yes playwright install-deps chromium && \
    apt autoremove -y nodejs && \
    rm -rf /root/.npm && \
    rm -rf /var/lib/apt/lists/* && \
    rm -rf $PLAYWRIGHT_BROWSERS_PATH/ffmpeg* && \
    rm -rf $PLAYWRIGHT_BROWSERS_PATH/firefox* && \
    rm -rf $PLAYWRIGHT_BROWSERS_PATH/webkit*

COPY docker/Spartan-Variable.ttf /usr/local/share/fonts/Spartan-Variable.ttf
RUN fc-cache -f && rm -rf /var/cache/*

RUN groupadd -r -g 999 opengraph && \
    useradd -u 999 -r -g 999 opengraph

COPY --chown=root:root target/opengraph.jar /opt/opengraph/opengraph.jar

USER opengraph
WORKDIR /tmp

ENV PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD 1
ENV PLAYWRIGHT_SKIP_BROWSER_GC 1

STOPSIGNAL 15
CMD ["java", "-jar", "/opt/opengraph/opengraph.jar"]
```

Note that this (perhaps importantly) means that we are a non-root user at runtime, in case process termination somehow expects that. However I'd then expect whatever call gets a permission denied to be logged, which isn't the case.

```java
import com.microsoft.playwright.*;

public class ExampleReproducible {

    private static final Logger LOGGER = LoggerFactory.getLogger(OpenGraph.class);

    public static void main(String[] args) throws InterruptedException {
        try (
            Playwright playwright = Playwright.create();
            Browser chromium = playwright.chromium().launch()
        ) {
            LOGGER.info("Loaded Chromium {}", chromium.version());
        } catch (Exception e) {
            throw new IllegalStateException("Failed initializing Chromium via Playwright!", e);
        }
        LOGGER.info("Sleeping...");
        Thread.sleep(60_000);
    }

}
```

**Describe the bug**

After this has ran, there are 2 zombie chrome processes left, despite running without errors and the close method of both Playwright and Browser being invoked.

```
$ docker run -it opengraph
Skipping browsers download because `PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD` env variable is set
08:09:09.720 [main] INFO OpenGraph - Loaded Chromium 101.0.4929.0
08:09:09.783 [main] INFO OpenGraph - Sleeping...
```

```
  PID TTY      STAT   TIME COMMAND
    1 pts/0    Ssl+   0:03 java -jar /opt/opengraph/opengraph.jar
  127 ?        Z      0:00 [chrome] <defunct>
  128 ?        Z      0:00 [chrome] <defunct>
  168 pts/1    Ss     0:00 bash
  176 pts/1    R+     0:00 ps ax
```

**Comment by Tristan971** [comment]
After a bit more experimentation, the following setup fixes the issue altogether:
```java
try (
    Playwright playwright = Playwright.create();
    Browser chromium = playwright.chromium().launch(new BrowserType.LaunchOptions().setArgs(List.of(
        "--single-process",
        "--no-zygote"
    )));
) {
    LOGGER.info("Loaded Chromium {}", chromium.version());
} catch (Exception e) {
    throw new IllegalStateException("Failed initializing Chromium via Playwright!", e);
}
```

It's not optimal but it does the job. Not sure how that interracts with `--no-sandbox` however. Docs of chrome flags (here https://peter.sh/experiments/chromium-command-line-switches/ ) suggests on linux the no-sandbox is also necessary, while there's also a `--no-zygote-sandbox` flag which supposedly allows sandboxing without the zygote...

**Comment by Tristan971** [comment]
Well this has other issues, notably that single-process doesn't work well with how playwright's driver is implemented, which is fair.

I can't figure out the whole of it, but by making sure to not throw within the Playwright loop (well, thread more specifically), which I was doing through not handling the TimeoutError inside it, then things are fine. Oh well.

**Comment by nikolasil** [comment]
I had the same issue with the zombie processes. After a lot of research I found out that by default docker uses an isolated pid namespace. So the pid's of host are different from the pid's of docker container processes.
Example: the pid 1 was for the java process inside the docker container. And of course the pid 1 was systemd in the host.

I added pid: "host" in the docker container. This way the pid's that the container's processes allocate are sequential to the pid's of the host. When the process loses his parent process and becomes a zombie process, his parent will become the process with pid 1 (systemd). systemd which is the master process waits all his children. So he waits for all those zombie processes and eventually stops them.

My solution for this issue was a "workaround". Cause the zombie processes problem still exists but are killed by the operating system. I don't know why they are left behind. Maybe a problem from browser(chromium and firefox as I tested) or Playwright not closing properly all the browser's threads.
